core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala

```diff
@@ -52,6 +52,13 @@ private[spark] sealed trait MapStatus {
    * partitionId of the task or taskContext.taskAttemptId is used.
    */
   def mapId: Long
+
+  /**
+   * Extra metadata for map status. This could be used by different ShuffleManager implementation
+   * to store information they need. For example, a Remote Shuffle Service ShuffleManager could
+   * store shuffle server information and let reducer task know where to fetch shuffle data.
+   */
+  def metadata: Option[Serializable]
```

Comment:
hmm...what's the relationship between SPARK-33114 and SPARK-25299? According to the JIRA description, SPARK-33114 seems to enhance the support for custom shuffle manager while SPARK-25299 only customize the storage with the default  `SortShuffleManager`.

So if we are only talking about SPARK-33114, adding `metadata` may be a good choice according to its own scenario. But if we bring in SPARK-25299 together (IIUC, what this PR is doing would also benefit SPARK-25299), I personally think we need a more general design here. For example, I'd prefer to redesign the `location` of `MapStatus` to make it be able to support different scenarios (e.g., Spark BlockManager, Spark external shuffle service, custom remote storage, etc. ) mentioned in SPARK-25299. And in this way, different scenarios would be able to reuse the existing features, e.g., decommission(which may update mapstatus location during runtime) and reuse the existing code paths, e.g., we don't need the extra `getAllMapOutputStatuses` and everything should be the same as what we already did during shuffle reading.

WDYT?

Reply:
@Ngone51 I agree with you that finishing the design laid out in SPARK-25299 would be much better.
This is why I opened https://github.com/apache/spark/pull/30763 as a copy of Matthew Cheah's original PR for SPARK-31801 (because he is busy with other projects) and kept it up-to-date several times with the master.

But it haven't got enough reviews and I wouldn't want to block @hiboyang further, https://github.com/apache/spark/pull/30004#issuecomment-782915655.

I am sure with your help we can complete SPARK-31801 an be on the road of SPARK-25299.
 
So next week I will do the conflict resolution and ping you when the PR is ready for review. Is this okay?

Reply:
Cool, thanks @attilapiros for keeping working on [SPARK-25299](https://issues.apache.org/jira/browse/SPARK-25299) while unblocking this PR. @Ngone51 [SPARK-33114](https://issues.apache.org/jira/browse/SPARK-33114) is a small change to support remote shuffle service/storage by adding a metadata object in MapStatus. It could be viewed as a subset of [SPARK-25299](https://issues.apache.org/jira/browse/SPARK-25299) 's work.

Reply:
> So next week I will do the conflict resolution and ping you when the PR is ready for review. Is this okay?

Sure, please.  @attilapiros

Reply:
@hiboyang Thanks for your explanation. I agree that #30763 is too big for review. But I think we can discuss there first to ensure we towards the same direction before we deep into details. And when we're on the same page, we can split the big PR into smaller pieces and start to co-work. Does it sound good to you?

Reply:
Yes, it will be great if we continue to discuss and make progress for #30763. Maybe we could have a sync up zoom meeting to discuss how to split the big work? There are several remote/disaggregated shuffle solutions from Spark community and different companies. We could walk through these solutions and come out ideas about the minimum set of changes to support them?

Reply:
I think we should continue the discussion based on the [SPIP](https://docs.google.com/document/d/1d6egnL6WHOwWZe8MWv3m8n4PToNacdx7n_0iMSWwhCQ/edit#) of SPARK-25299. I assume the SPIP of SPARK-25299 already passed since there're already some changes merged in, which means people already discussed other solutions previously. But sure, feel free to raise your concern if the current design still can't cover your use case.

And we actually know what we need to do (you can also check the sub-tasks of SPARK-25299). By further discussion, I mean, for the case like SPARK-31801, 1) we could split it into smaller pieces since its current scope is too big (but we need to take a look at the PR first) 2) the current implementation is still not finalized, we can discuss other ways for SPARK-31801.

Reply:
[SPIP of SPARK-25299](https://docs.google.com/document/d/1d6egnL6WHOwWZe8MWv3m8n4PToNacdx7n_0iMSWwhCQ/edit#) says "API proposed here sits behind the existing SortShuffleManager implementation of the ShuffleManager interface". While we could continue the work in that area (behind the existing SortShuffleManager), could we also discuss from another point of view like supporting shuffle manager implementations other than SortShuffleManager?

Reply:
Then, it'd be a totally different topic, right?  IIUC, SPARK-25299 could also benefit custom shuffle manager if SPARK-25299(custom storage) is pluggable with a custom shuffle manager. Ideally, a custom shuffle manager should be able to plugin in different storages. We might not think deeply about how to support custom shuffle managers when working on SPARK-25299 but need to keep in mind the "pluggable". After SPARK-25299 completed, then, we can start to enhance the support for custom shuffle managers.

That being said, I think you are still free to raise a separate discussion on supporting the custom shuffle manager. For example, what are the shortcomings of the current framework, and what should be improved...I just wonder the community may not have enough bandwidth to work on these two significant projects concurently.

Reply:
Yes, in this case, agree with you that this PR is a different topic from [SPARK-25299](https://issues.apache.org/jira/browse/SPARK-25299), and the community not have enough bandwidth to work on these two significant projects concurrently. This PR is adding a simple change to make it possible that different custom shuffle managers could add their own metadata inside the MapStatus object. Could we proceed to review this PR then?

Reply:
I can see how this PR would work. But even if we merged this PR, the ongoing SPARK-31801 may override your change before the release. `getAllMapOutputStatusMetadata` isn't an API, so you may suffer the breaking change. We're discussing a general way in SPARK-31801 to provide the API for users. IIUC, it'd also benefit your case.

WDYT?

Reply:
I feel [SPARK-31801](https://issues.apache.org/jira/browse/SPARK-31801) is very big, and may take very long time to finish. Could we merge this PR first?

If [SPARK-31801](https://issues.apache.org/jira/browse/SPARK-31801) find a better way to support it and break `getAllMapOutputStatusMetadata`, it is actually a good thing :) We could have multiple iterations. This PR is the first iteration with very small change. [SPARK-31801](https://issues.apache.org/jira/browse/SPARK-31801) is the iteration after that. The latter does not need to block the former one.

Reply:
I don't think we should develop like this way...As you mentioned above, SPARK-33114 can be considered as a subtask of SPARK-25299. So how can we consider this PR as a first iteration when SPARK-25299 is still under discussion and development, especially when people haven't reached an agreement on the solution and has a possible alternative solution at the same time? Also, I think the custom shuffle manager isn't officially supported by Spark because the `ShuffleManager` interface is private. So it doesn't make sense for Spark to add an internal API for un-official use cases if there's no strong reason.

SPARK-31801 is surely big. But as I mentioned early, we can split it. When the solution is finalized, we can start with refactoring `MapStatus` first. I think it would be a much smaller task and be enough for your case. And then, we'll start the remaining work(e.g. use the new `MapStatus` where it was referenced) but you don't care.

I understand you have paid a lot of effort into this work, and sorry we can not get it in fast. And, unfortunately, I don't have the permission to merge. You could persuade committers to merge the PR if you insist on it.

Reply:
Yeah, it is also good idea if we could split [SPARK-31801](https://issues.apache.org/jira/browse/SPARK-31801) and start with refactoring `MapStatus` first. Do you or the community get ideas about how to split [SPARK-31801](https://issues.apache.org/jira/browse/SPARK-31801)?

Reply:
We're still discussing the solution in #30763. So I can't tell you the concrete split plan. But, I think, we'd be able to start with refactoring `MapStatus` either way.